# Exploring Student Data

There are some important tasks with looking for answering questions below:
- How are students performing in their math classes?
- What do students’ parents do for work?
- How often are students absent from school?

Data citation:

Dua, D. and Graff, C. (2019). UCI Machine Learning Repository [http://archive.ics.uci.edu/ml]. Irvine, CA: University of California, School of Information and Computer Science. http://archive.ics.uci.edu/ml/datasets/Student+Performance

Paulo Cortez, University of Minho, Guimarães, Portugal, http://www3.dsi.uminho.pt/pcortez

In [1]:
# load libraries and setting options
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option('display.max_columns', None)

In [2]:
# import data 
students = pd.read_csv('student-por.csv', sep=';', encoding='utf-8')

In [3]:
# Print first few rows of data
students

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
644     MS   F   19       R     GT3       T     2     3  services     other   
645     MS   F   18       U     LE3       T     3     1   teacher  services   
646     MS   F   18       U     GT3       T     1     1     other     other   
647     MS   M   17       U     LE3       T     3     1  services  services   
648     MS   M   18       R     LE3       T     3     2  services     other   

     reason guardian  traveltime  studytime  failures schoolsup famsup paid  \
0    course   mother           2          2         0       yes     no   no   
1    course   father           1          2         0        no    yes   no   
2     other   mother           1          2         0       yes     no   no   
3      home   mother           1          3         0        no    yes   no   
4      home   father           1          2         0        no    yes   no   
..      ...      ...         ...        ...       ...       ...    ...  ...   
644  course   mother           1          3         1        no     no   no   
645  course   mother           1          2         0        no    yes   no   
646  course   mother           2          2         0        no     no   no   
647  course   mother           2          1         0        no     no   no   
648  course   mother           3          1         0        no     no   no   

    activities nursery higher internet romantic  famrel  freetime  goout  \
0           no     yes    yes       no       no       4         3      4   
1           no      no    yes      yes       no       5         3      3   
2           no     yes    yes      yes       no       4         3      2   
3          yes     yes    yes      yes      yes       3         2      2   
4           no     yes    yes       no       no       4         3      2   
..         ...     ...    ...      ...      ...     ...       ...    ...   
644        yes      no    yes      yes       no       5         4      2   
645         no     yes    yes      yes       no       4         3      4   
646        yes     yes    yes       no       no       1         1      1   
647         no      no    yes      yes       no       2         4      5   
648         no      no    yes      yes       no       4         4      1   

     Dalc  Walc  health  absences  G1  G2  G3  
0       1     1       3         4   0  11  11  
1       1     1       3         2   9  11  11  
2       2     3       3         6  12  13  12  
3       1     1       5         0  14  14  14  
4       1     2       5         0  11  13  13  
..    ...   ...     ...       ...  ..  ..  ..  
644     1     2       5         4  10  11  10  
645     1     1       1         4  15  15  16  
646     1     1       5         6  11  12   9  
647     3     4       2         6  10  10  10  
648     3     4       5         4  10  11  11  

[649 rows x 33 columns]

In [4]:
students.shape

(649, 33)

In [8]:
# print out summary statistics for all features in the dataset
students.describe(include='all')

school  sex         age address famsize Pstatus        Medu  \
count     649  649  649.000000     649     649     649  649.000000   
unique      2    2         NaN       2       2       2         NaN   
top        GP    F         NaN       U     GT3       T         NaN   
freq      423  383         NaN     452     457     569         NaN   
mean      NaN  NaN   16.744222     NaN     NaN     NaN    2.514638   
std       NaN  NaN    1.218138     NaN     NaN     NaN    1.134552   
min       NaN  NaN   15.000000     NaN     NaN     NaN    0.000000   
25%       NaN  NaN   16.000000     NaN     NaN     NaN    2.000000   
50%       NaN  NaN   17.000000     NaN     NaN     NaN    2.000000   
75%       NaN  NaN   18.000000     NaN     NaN     NaN    4.000000   
max       NaN  NaN   22.000000     NaN     NaN     NaN    4.000000   

              Fedu   Mjob   Fjob  reason guardian  traveltime   studytime  \
count   649.000000    649    649     649      649  649.000000  649.000000   
unique         NaN      5      5       4        3         NaN         NaN   
top            NaN  other  other  course   mother         NaN         NaN   
freq           NaN    258    367     285      455         NaN         NaN   
mean      2.306626    NaN    NaN     NaN      NaN    1.568567    1.930663   
std       1.099931    NaN    NaN     NaN      NaN    0.748660    0.829510   
min       0.000000    NaN    NaN     NaN      NaN    1.000000    1.000000   
25%       1.000000    NaN    NaN     NaN      NaN    1.000000    1.000000   
50%       2.000000    NaN    NaN     NaN      NaN    1.000000    2.000000   
75%       3.000000    NaN    NaN     NaN      NaN    2.000000    2.000000   
max       4.000000    NaN    NaN     NaN      NaN    4.000000    4.000000   

          failures schoolsup famsup paid activities nursery higher internet  \
count   649.000000       649    649  649        649     649    649      649   
unique         NaN         2      2    2          2       2      2        2   
top            NaN        no    yes   no         no     yes    yes      yes   
freq           NaN       581    398  610        334     521    580      498   
mean      0.221880       NaN    NaN  NaN        NaN     NaN    NaN      NaN   
std       0.593235       NaN    NaN  NaN        NaN     NaN    NaN      NaN   
min       0.000000       NaN    NaN  NaN        NaN     NaN    NaN      NaN   
25%       0.000000       NaN    NaN  NaN        NaN     NaN    NaN      NaN   
50%       0.000000       NaN    NaN  NaN        NaN     NaN    NaN      NaN   
75%       0.000000       NaN    NaN  NaN        NaN     NaN    NaN      NaN   
max       3.000000       NaN    NaN  NaN        NaN     NaN    NaN      NaN   

       romantic      famrel    freetime       goout        Dalc        Walc  \
count       649  649.000000  649.000000  649.000000  649.000000  649.000000   
unique        2         NaN         NaN         NaN         NaN         NaN   
top          no         NaN         NaN         NaN         NaN         NaN   
freq        410         NaN         NaN         NaN         NaN         NaN   
mean        NaN    3.930663    3.180277    3.184900    1.502311    2.280431   
std         NaN    0.955717    1.051093    1.175766    0.924834    1.284380   
min         NaN    1.000000    1.000000    1.000000    1.000000    1.000000   
25%         NaN    4.000000    3.000000    2.000000    1.000000    1.000000   
50%         NaN    4.000000    3.000000    3.000000    1.000000    2.000000   
75%         NaN    5.000000    4.000000    4.000000    2.000000    3.000000   
max         NaN    5.000000    5.000000    5.000000    5.000000    5.000000   

            health    absences          G1          G2          G3  
count   649.000000  649.000000  649.000000  649.000000  649.000000  
unique         NaN         NaN         NaN         NaN         NaN  
top            NaN         NaN         NaN         NaN         NaN  
freq           NaN         NaN         NaN         NaN         NaN 

In [ ]:
# Calculate and print the mean value of G3_grade
mean_G3 = students['G3'].mean()
print(f"The mean value of G3_grade is: {mean_G3:.2f}")

The mean value of G3_grade is: 11.91


In [10]:
# Calculate and print the median value of G3_grade
median_G3 = students['G3'].median()
print(f"The median value of G3_grade is: {median_G3:.2f}")

The median value of G3_grade is: 12.00


In [14]:
# Calculate and print the mode value of G3_grade
mode_G3 = students['G3'].mode().iloc[0]
print(f"The mode value of G3_grade is: {mode_G3:.2f}")

The mode value of G3_grade is: 11.00


In [ ]:
# Calculate and print the range of G3_grade
range_G3 = students['G3'].max() - students['G3'].min()
print(f"The range of G3_grade is: {range_G3:.2f}")

The range of G3_grade is: 19.00


In [16]:
# Calculate and print the standard deviation of G3_grade
std_G3 = students['G3'].std()
print(f"The standart deviation of G3_grade is: {std_G3:.2f}")

The standart deviation of G3_grade is: 3.23


About two thirds of students are earning a grade G3 between 8.68 (calculated as 11.91-3.23) and 15.14 (calculated as 11.91+3.23)

In [21]:
# Calculate and print the mean absolute deviation of G3_grade
mad_G3 = (students['G3'] - students['G3'].mean()).abs().mean()
print(f"The mean absolute deviation of G3_grade is: {mad_G3:.2f}")

The mean absolute deviation of G3_grade is: 2.41
